### PR TITLE
feat(headless SSR): support both search and listing solution types

### PR DESCRIPTION
**Pull request by y-lakhdar**
## TL;DR

Enhance the commerce engine to support both search and listing solution types, implementing type inference and conditional types to ensure the correct controllers are instantiated based on the solution type.

### **Solution Type Support**:
The commerce engine now supports both search and listing solution types. This is particularly important when building sub-controllers (e.g., summary, facets, sort, etc.). Since, in SSR, users do not manually build their controllers because it is done under the hood (in the SSR utils), the SSR utils need to know which function to call (e.g., `buildProductListing()` or `buildSearch()`).

### **Conditional Controller Instantiation**: 
Based on the specified solution type, the engine now conditionally instantiates controllers. This mechanism ensures that only relevant controllers are active for a given solution type, optimizing the engine’s performance and resource usage by avoiding the creation of unnecessary controllers.

#### Example
For instance, if you create an engine definition containing a Query Summary but want the QuerySummary controller to not be created on listing pages, you can opt out for this solution type. Otherwise, it will be built like other controllers.

```ts
const engineDefinition = defineCommerceEngine({
  configuration: configuration,
  controllers: {
    searchbox: defineStandaloneSearchBox({
      options: {redirectionUrl: '/search'},
    }),
    summary: defineQuerySummary({listing: false}), // opting out for listing context
    // ... other controllers
  },
});
```

### **Type Inference Enhancements**:
Since Conditional Controller Instantiation is a runtime logic, I had to update the typing to reflect that change so that the same logic could also be available statically with TypeScript. This means that if you opt out from a solution type during the engine definition (like in the previous example), you should not have the controller available when using code completion.

These enhancements facilitate the accurate determination of controller types based on the solution type, streamlining the setup process for developers.

#### Example
Using the same engine definition from the previous example, and since we have opted out from query summary in the listing context, you will notice that you get an error if you try to get that controller in the listing context.
image

### **Duplicate Types and Methods**: 
Due to compatibility issues with existing types used by the non-commerce SSR package, some types and methods have been duplicated and adapted to fit the solution type strategy.

https://coveord.atlassian.net/browse/KIT-3394

### CODE REVIEW

File: packages/headless/src/app/commerce-ssr-engine/common.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+}): InferControllersMapFromDefinition<
+  TControllerDefinitionsMap,
+  TSolutionType
+> {
+  const controllerMap = mapObject(definitionsMap, (definition, key) => {
+    // TODO: find a way to make it dynamic
+    const searchDisabled =
+      'search' in definition &&
+      definition['search'] === false &&
+      solutionType === SolutionType['search'];
```
Comment (alexprudhomme) [inline packages/headless/src/app/commerce-ssr-engine/common.ts:63]: How is search 'disabled' if `'search'` is in the definition & the solutionType is `'search'` ?

File: packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts [hunk 2]
```diff
@@ -128,97 +145,129 @@ export function defineCommerceEngine<
...
+        },
+      }
+    );
   return {
-    build,
-    fetchStaticState,
-    hydrateStaticState,
-    setNavigatorContextProvider,
+    listingEngineDefinition: {
+      build: buildFactory(SolutionType.listing),
```
Comment (alexprudhomme) [inline packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts:261]: What is the `build` function used for ? I don't think it is used in the sample.

File: packages/headless/src/app/commerce-ssr-engine/common.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    const listingDisabled =
+      'listing' in definition &&
+      definition['listing'] === false &&
+      solutionType === SolutionType['listing'];
+
+    if (searchDisabled || listingDisabled) {
+      return null;
+    }
```
Comment (alexprudhomme) [inline packages/headless/src/app/commerce-ssr-engine/common.ts:72]: What happens here ? The definition will have a null controller ?

File: packages/headless/src/app/search-engine/search-engine.ssr.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
 /**
  * Utility functions to be used for Server Side Rendering.
  */
-import {UnknownAction} from '@reduxjs/toolkit';
+import {AnyAction, UnknownAction} from '@reduxjs/toolkit';
```
Comment (alexprudhomme) [inline packages/headless/src/app/search-engine/search-engine.ssr.ts:4]: Same thing here about AnyAction. It seem to be deprecated and `UnknownAction ` is the exact same thing in the source code.

File: packages/samples/headless-ssr-commerce/app/_lib/commerce-engine-config.ts [hunk 5]
```diff
@@ -4,33 +4,57 @@ import {
...
-
 export default {
   configuration: {
-    ...configuration,
+    accessToken: 'xxc481d5de-16cb-4290-bd78-45345319d94c',
+    organizationId: 'barcasportsmcy01fvu',
+    environment: 'dev',
+    analytics: {
+      trackingId: 'sports',
+    },
```
Comment (alexprudhomme) [inline packages/samples/headless-ssr-commerce/app/_lib/commerce-engine-config.ts:21]: Why are we on barca. Can't we use searchuisamples now ?

File: packages/headless/src/app/commerce-ssr-engine/types/core-engine.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+export type EngineDefinitionOptions<
+  TOptions extends {configuration: EngineConfiguration},
+  TControllers extends ControllerDefinitionsMap<
+    CoreEngine | CoreEngineNext,
+    Controller
+  >,
+> = TOptions & {
+  /**
+   * The controllers to initialize with the search engine.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/app/commerce-ssr-engine/types/core-engine.ts:25]:
```suggestion
   * The controllers to initialize with the commerce engine.
```

File: packages/samples/headless-ssr-commerce/app/_lib/commerce-engine-config.ts (hunk 5 above)
Comment (fbeaudoincoveo) [reply packages/samples/headless-ssr-commerce/app/_lib/commerce-engine-config.ts:21]: I think we could use the getSampleCommerceConfiguration, but as long as there isn't a context controller, we'll have to override the view.url in the config to target an existing product listing page configuration

File: packages/samples/headless-ssr-commerce/app/listing/page.tsx [hunk 7]
```diff
@@ -1,15 +1,27 @@
...
 
 /**
  * This file defines a List component that uses the Coveo Headless SSR commerce library to manage its state.
  *
  * The Listing function is the entry point for server-side rendering (SSR).
  */
 export default async function Listing() {
-  const staticState = await fetchStaticState();
+  // Sets the navigator context provider to use the newly created `navigatorContext` before fetching the app static state
+  const navigatorContext = new NextJsNavigatorContext(headers());
```
Comment (fbeaudoincoveo) [inline packages/samples/headless-ssr-commerce/app/listing/page.tsx:13]: I asked the same question in one of @alexprudhomme 's PR, so feel free to ignore if you've already answered there, but I wonder why we're doing this here instead of passing the navigator context provider when we define the engine.

**Review by fbeaudoincoveo — APPROVED** [review]

**Review by alexprudhomme — APPROVED** [review]
